genesis/genesis.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2015-2017 Monax Industries Limited.
+// This file is part of the Monax platform (Monax)
+
+// Monax is free software: you can use, redistribute it and/or modify
+// it only under the terms of the GNU General Public License, version
+// 3, as published by the Free Software Foundation.
+
+// Monax is distributed WITHOUT ANY WARRANTY pursuant to
+// the terms of the Gnu General Public Licence, version 3, including
+// (but not limited to) Clause 15 thereof. See the text of the
+// GNU General Public License, version 3 for full terms.
+
+// You should have received a copy of the GNU General Public License,
+// version 3, with Monax.  If not, see <http://www.gnu.org/licenses/>.
+
+package genesis
+
+import (
+	"bytes"
+	"encoding/json"
+	"time"
+
+	ptypes "github.com/eris-ltd/eris-db/permission/types"
+	wire "github.com/tendermint/go-wire"
+)
+
+// MakeGenesisDocFromAccounts takes a chainName and a slice of pointers to GenesisAccount,
+// and a slice of pointers to GenesisValidator to construct a GenesisDoc, or returns an error on
+// failure.  In particular MakeGenesisDocFromAccount uses the local time as a
+// timestamp for the GenesisDoc.
+func MakeGenesisDocFromAccounts(chainName string, accounts []*GenesisAccount,
+	validators []*GenesisValidator) (GenesisDoc, error) {
+
+	// TODO: assert valid accounts and validators
+	// TODO: [ben] expose setting global permissions
+	globalPermissions := ptypes.DefaultAccountPermissions.Clone()
```

Comment:
Yes, but is a struct `AccountPermissions` of a struct `BasePermissions` and a slice of `strings`; I believe - but not sure - that the struct `BasePermissions` is treated as a pointer essentially; but definitely the `Roles []string` is an effective pointer;

secondly, genesis is currently constructed with `Params` as a pointer to a `GenesisParams` which is a pointer to an `AccountPermissions`;  I dont particularly like this construction; but I didnt want to end up with a `GenesisDoc` that didnt own its own data;

Arguably this does read very defensive